File: 1025-minimum-cost-for-tickets/minimum-cost-for-tickets.py
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        n=len(days)
        ma=days[n-1]
        ticket=[1,7,30]
        dp=[ 1000000007 for i in range(ma+1)]
        i=0
        dp[0]=0
        
        for i in range(1,-1,-1):
            if costs[i]>costs[i+1]:
                costs[i]=costs[i+1]
        for k in range(1,ma+1):
            if k<days[i]:
                dp[k]=dp[k-1]
            else:
                for l in range(3):
                    dp[k]=min(dp[k],dp[max(0,(k-ticket[l]))]+costs[l])
                i+=1
            
        print(dp)
        return dp[ma]







        
        

File: 1025-minimum-cost-for-tickets/minimum-cost-for-tickets.java
class Solution {
    HashSet<Integer> isTravelNeeded = new HashSet<>();

    private int solve(int[] dp, int lastDay, int[] costs, int currDay) {
        // If we have iterated over travel days, return 0.
        if (currDay > lastDay) {
            return 0;
        }

        // If we don't need to travel on this day, move on to next day.
        if (!isTravelNeeded.contains(currDay)) {
            return solve(dp, lastDay, costs, currDay + 1);
        }

        // If already calculated, return from here with the stored answer.
        if (dp[currDay] != -1) {
            return dp[currDay];
        }

        int oneDay = costs[0] + solve(dp, lastDay, costs, currDay + 1);
        int sevenDay = costs[1] + solve(dp, lastDay, costs, currDay + 7);
        int thirtyDay = costs[2] + solve(dp, lastDay, costs, currDay + 30);

        // Store the cost with the minimum of the three options.
        return dp[currDay] = Math.min(oneDay, Math.min(sevenDay, thirtyDay));
    }

    public int mincostTickets(int[] days, int[] costs) {
        // The last day on which we need to travel.
        int lastDay = days[days.length - 1];
        int dp[] = new int[lastDay + 1];
        Arrays.fill(dp, -1);

        // Mark the days on which we need to travel.
        for (int day : days) {
            isTravelNeeded.add(day);
        }

        return solve(dp, lastDay, costs, 1);
    }
}